### PR TITLE
Fix #282 - Refactor archives.py and make `aqt tool` take --version and --arch as optionals

### LINKED ISSUE #282: Enhancement Request: Change `aqt tool` to not require version and arch (--latest)
I can see that `tools_ifw` always breaks because Qt will remove any previous versions (there's only one). It seems we could just parse the Updates.xml and get the latest one if the version isn't actually supplied.

Also, it seems like error reporting needs to be improved. Even if the tool isn't found, the "Installation finished" is issued.

I will try to provide a PR for review

### PR DESCRIPTION
**Pull request by jmarrec**
- Use packaging.version instead of semantic_version
    - That way you could pass --version 4, it would allow building a version, and it would match any version that starts with 4. See the helper I added, and the unit tests that show the behavior clearly  
- Provide `__str__`, `__repr__` for classes for easier debugging
- Use `posixpath` instead of string manipulation to make urls
- Add type hints were I touched thing
   -  TODO: should add mypi as a checker

Edit: it's likely that some of the workflows will fail, and that's fine. I only ran the unit test locally + try `aqt tool`, haven't tried Qt. We can address problems as they are caught

### CODE REVIEW

File: aqt/helper.py [hunk 1]
```diff
@@ -205,6 +207,17 @@ def handle(self, record):
...
 
+def satifiesVersion(requestedVersion: Version, candidateVersion: Version):
+    """
+    This will return true if the candidateVersion matches the requestedVersion
+    It will only compare the number of components of the candidateVersion
+    """
+    for (a, b) in zip(requestedVersion.release, candidateVersion.release):
+        if a != b:
+            return False
+    return True
```
Comment (jmarrec) [inline aqt/helper.py:218]: New helper.

v.release returns a tuple

```
v1 = Version('4.0') => v1.release = (4, 0)
v2 = Version('4.0.1') => v2.release = (4, 0, 1)

zip:
  4 == 4
  0 == 0
  
  => it matches
```

File: aqt/installer.py [hunk 2]
```diff
@@ -427,13 +427,14 @@ def run_tool(self, args):
...
-                os_name,
-                tool_name,
-                version,
-                arch,
-                base,
+                os_name=os_name,
+                tool_name=tool_name,
+                base=base,
+                version_str=version,
+                arch=arch,
```
Comment (jmarrec) [inline aqt/installer.py:437]: Avoid positional args, so we can reorder without problems

File: aqt/installer.py [hunk 3]
```diff
@@ -635,10 +636,17 @@ def _create_parser(self):
...
+            type=str,
+            help=('Tool version in the format of "4.1.2". '
+                  'If not supplied, defaults to latest found'),
         )
+
         tools_parser.add_argument(
-            "arch", help="Name of full tool name such as qt.tools.ifw.31"
+            '-a', '--arch',
+            type=str,
+            help='Name of full tool name such as "qt.tools.ifw.31"'
```
Comment (jmarrec) [inline aqt/installer.py:649]: Make these two optionals

File: aqt/installer.py [hunk 4]
```diff
@@ -676,6 +684,8 @@ def run(self, arg=None):
         return result
 
     def call_installer(self, qt_archives, base_dir, sevenzip, keep):
+        if not qt_archives.get_archives():
+            raise NoPackageFound("Couldn't find any archives")
```
Comment (jmarrec) [inline aqt/installer.py:688]: I will be raising earlier in `ToolsArchives` with a better error message, but I think it's fair to have this as a fallback...

File: tests/test_helper.py [hunk 5]
```diff
@@ -100,3 +102,26 @@ def test_helper_downloadBinary_sha256(tmp_path, monkeypatch):
...
+    ],
+)
+def test_satifiesVersion(requestedVersion, candidateVersion, expected):
+    v1 = packaging.version.parse(requestedVersion)
+    v2 = packaging.version.parse(candidateVersion)
+    matches = helper.satifiesVersion(v1, v2)
+    if expected:
+        assert matches, f"Expected {candidateVersion} to satify {requestedVersion}"
+    else:
+        assert not matches, f"Expected {candidateVersion} to NOT match {requestedVersion}"
```
Comment (jmarrec) [inline tests/test_helper.py:127]: Test shows the way satisfiesVersion works

File: aqt/archives.py [hunk 6]
```diff
@@ -36,18 +38,40 @@ def __init__(self, version, target, arch, os_name):
         self.arch = arch
         self.os_name = os_name
 
+    def __str__(self):
+        print(f'TargetConfig(version={self.version}, target={self.target}, '
+              f'arch={self.arch}, os_name={self.os_name}')
+
+    def __repr__(self):
+        print(f'({self.version}, {self.target}, {self.arch}, {self.os_name})')
+
```
Comment (jmarrec) [inline aqt/archives.py:47]: Adding __repr__ and __str__ for easier debugging: you can just print objects (__str__) or lists of objects (__repr__ is used)

File: aqt/archives.py [hunk 7]
```diff
@@ -81,7 +106,9 @@ def _get_archives(self):
         # Get packages index
         if self.version.major == 6 and self.target == "android":
             arch_ext = ["_armv7/", "_x86/", "_x86_64/", "_arm64_v8a/"]
-        elif self.version in SimpleSpec(">=5.13.0,<6.0") and self.target == "desktop":
+        elif ((self.target == 'desktop') and
+              (self.version >= Version('5.13.0')) and
+              (self.version < Version('6.0.0'))):
```
Comment (jmarrec) [inline aqt/archives.py:111]: port from semantic_version to packaging.version, an example

File: aqt/archives.py [hunk 8]
```diff
@@ -92,10 +119,10 @@ def _get_archives(self):
                 self.target,
                 self.version.major,
                 self.version.minor,
-                self.version.patch,
+                self.version.micro,
                 ext,
             )
-            update_xml_url = "{0}{1}Updates.xml".format(self.base, archive_path)
+            update_xml_url = posixpath.join(self.base, archive_path, "Updates.xml")
```
Comment (jmarrec) [inline aqt/archives.py:125]: prefer posixpath to avoid any problems with `/` separators and co

File: aqt/archives.py [hunk 9]
```diff
@@ -199,16 +227,20 @@ def _get_archives(self):
...
         target_packages.append(
             "qt.{0}{1}{2}.{3}".format(
-                self.version.major, self.version.minor, self.version.patch, self.arch
+                self.version.major, self.version.minor, self.version.micro, self.arch
             )
         )
         target_packages.extend(self.mod_list)
+        self.logger.debug('QtArchives::_get_archives')
+        self.logger.debug(f'{update_xml_url=}')
+        self.logger.debug(f'{target_packages=}')
```
Comment (jmarrec) [inline aqt/archives.py:242]: perhaps shoudl remove, dunno

File: aqt/archives.py [hunk 10]
```diff
@@ -36,18 +38,40 @@ def __init__(self, version, target, arch, os_name):
...
-    def __init__(self, name, archive_url, archive, package_desc, hashurl):
+    def __init__(self, name: str , archive_url: str, archive: str,
+                 package_desc: str, hashurl: str,
+                 version: Optional[Version] = None):
         self.name = name
         self.url = archive_url
         self.archive = archive
         self.desc = package_desc
         self.hashurl = hashurl
+        self.version = version
```
Comment (jmarrec) [inline aqt/archives.py:62]: I store a version on the package. 
In the case of tools, there might be several matches. We want to filter on the max version to return a single

File: aqt/archives.py [hunk 11]
```diff
@@ -380,40 +441,69 @@ def _parse_update_xml(self, archive_url, target_packages):
...
+            full_version = packageupdate.find("Version").text
+            # keep only '4.1.1' in '4.1.1-20210526113'
+            parsed_version = packaging.version.parse(full_version.split('-')[0])
+            versions.append((name, str(parsed_version)))
+
+            # Filter out on version first
+            parsed_version = packaging.version.parse(full_version.split('-')[0])
+
+            if self.version:
+                if not satifiesVersion(self.version, parsed_version):
```
Comment (jmarrec) [inline aqt/archives.py:458]: This is the main change.

File: aqt/archives.py [hunk 12]
```diff
@@ -380,40 +441,69 @@ def _parse_update_xml(self, archive_url, target_packages):
...
-                        QtPackage(name, package_url, archive, package_desc, hashurl)
-                    )
+
+            # Filter out on arch next
+            if self.arch and name != self.arch:
+                self.logger.warning(
+                    f'Arch of {name} is different from requested '
+                    f'arch {self.arch} -- skip.'
+                )
+                continue
```
Comment (jmarrec) [inline aqt/archives.py:472]: Filter on arch next, if arch is given

File: aqt/archives.py [hunk 13]
```diff
@@ -380,40 +441,79 @@ def _parse_update_xml(self, archive_url, target_packages):
...
+            self.logger.debug(f'ToolArchives: {self.archives}')
+        else:
+            msg = f"{self} did not match any available tools. "
+            if versions:
+                msg += 'Available tools\n'
+                msg += f'[(version, arch)] = {versions}'
+            else:
+                url = posixpath.join(archive_url, "Updates.xml")
+                msg += f"Zero tools where found to be available in {url}"
+            raise NoPackageFound(msg)
```
Comment (jmarrec) [inline aqt/archives.py:506]: If none matches, raise with a helpful message

```
Arch of qt.tools.ifw.41 is different from requested arch qt.tools.ifw.40 -- skip.
Traceback (most recent call last):
 [...]
aqt.exceptions.NoPackageFound: ToolArchives(tool_name=tools_ifw, version=4.1, arch=qt.tools.ifw.40) did not match any available tools. Available tools
[(version, arch)] = [('qt.tools.ifw.41', '4.1.1')]

```

File: aqt/archives.py [hunk 14]
```diff
@@ -380,40 +441,79 @@ def _parse_update_xml(self, archive_url, target_packages):
...
+
+        if len(self.archives) > 1:
+            self.logger.info(
+                f"Found {len(self.archives)} potential archives. "
+                "Returning the max.")
+            archive_max = max(self.archives,
+                              key=lambda package: package.version
+                              if package.version
+                              else packaging.version.Version('0'))
+            self.archives = [archive_max]
```
Comment (jmarrec) [inline aqt/archives.py:516]: If more than one, only keep the max one.

**Review by jmarrec — COMMENTED** [review]
added review comments as a walkthrough of my changes

File: aqt/installer.py (hunk 4 above)
Comment (jmarrec) [reply aqt/installer.py:694]: Hum this addition is this resulting in the Azure failures: https://dev.azure.com/miurahr/github/_build/results?buildId=4250&view=logs&j=e9afaa34-1a0f-534e-78fc-fae528ccd915&t=55eb6c4a-c635-53c7-061a-ddccec759601&l=24

```
python -m aqt src 6.1.0 linux desktop --archives qt
```

**Comment by miurahr** [comment]
You are doing several different things at once.
Could you split PR?

**Comment by jmarrec** [comment]
> You are doing several different things at once.
> Could you split PR?

Can you clarify what you would want separated please?

The majority of the changes I am doing in other areas than "tool" is because I changed `semantic_version` to `packaging.version`. Which was something I actually needed, since semantic_version will refuse a string like '4.0' and throw. 
850a369df29ceedc5426d386b648fdb81cd9675a was perhaps ill-advised, I commented it out for now, I can revert it completely depending on CI outcome

**Comment by ddalcino** [comment]
> The majority of the changes I am doing in other areas than "tool" is because I changed `semantic_version` to `packaging.version`. Which was something I actually needed, since semantic_version will refuse a string like '4.0' and throw.

I think you might be better off using `semantic_version` instead of `packaging.version`. It already does everything you're trying to accomplish with `helper.satisfiesVersion`, and it also parses versions like "4.1.1-202105261130" without any problem.

If you want to accept strings like '4.0' as user input (for example, the user wants version '4.0' but doesn't care about the rest of the version number), you can use semantic_version.SimpleSpec. For example:

```python
Version('4.1.1-202105261130') in SimpleSpec('4.1')     # => True
Version('4.1.1-202105261130') in SimpleSpec('4.0')     # => False
Version('4.1.1-202105261130') in SimpleSpec('4')       # => True
```

**Comment by jmarrec** [comment]
@ddalcino packaging is more likely to be installed already since it's used by setuptools though.

I didn't realize SimpleSpec could work because I also thought I would support versions like '4.x' originally and `SimpleSpec('4.x')` throws (later realized LegacyVersion doesn't support '.release' method).

I don't really care either way as long as partial versions are working.

**Comment by ddalcino** [comment]
This project only recently switched from `packaging.version` to `semantic_version.Version` a couple of weeks ago; see #260 and #263. IIRC, we decided that it does a better job parsing version strings, and it's a more capable library in general. If you have a good reason as to why we should switch back, I would love to hear it. Maybe that could be a separate issue/PR?

> @ddalcino packaging is more likely to be installed already since it's used by setuptools though.

`semantic_version` is listed as a required dependency in `setup.cfg`, so it will be installed automatically when users install `aqt`. Yes, it is an additional dependency, and it weighs in at 50K disk space according to `du`. It's also popular enough that many users will already have it installed.

> I didn't realize SimpleSpec could work because I also thought I would support versions like '4.x' originally and SimpleSpec('4.x') throws (later realized LegacyVersion doesn't support '.release' method).

If you're trying to make `aqt tool` accept input like that, I really think you should at least take a look at SimpleSpec at https://pypi.org/project/semantic-version/. SimpleSpec supports versions like '4.*' and '>=4' by default:

```python
Version('4.1.1-202105261130') in SimpleSpec('4.*')    # => True
Version('4.1.1-202105261130') in SimpleSpec('>=4')    # => True
Version('4.1.1-202105261130') in SimpleSpec('>=4.2')  # => False
```

I suspect that attempting to re-create this kind of behavior using `packaging.version` and your own code will be outside the scope of this PR.

**Comment by jmarrec** [comment]
Ok I see your point, especially given the (recent) history. I will revert those changes back to semantic_version when I get some free time then

**Comment by jmarrec** [comment]
I will close this PR, and break it into several chunks

**Comment by jmarrec** [comment]
First PR: https://github.com/miurahr/aqtinstall/pull/287

**Pull request closed without merging**